tests/LinAlg/vectorTest.hpp

```diff
@@ -1527,19 +1537,151 @@ class VectorTests : public TestBase
     return reduceReturn(fail, &x);
   }
 
-protected:
-  // Interface to methods specific to vector implementation
-  virtual void setLocalElement(hiop::hiopVector* x, local_ordinal_type i, real_type val) = 0;
-  virtual real_type getLocalElement(const hiop::hiopVector* x, local_ordinal_type i) = 0;
-  virtual local_ordinal_type getLocalSize(const hiop::hiopVector* x) = 0;
-  virtual real_type* getLocalData(hiop::hiopVector* x) = 0;
-  virtual int verifyAnswer(hiop::hiopVector* x, real_type answer) = 0;
+  /// Method to set vector _x_ element _i_ to _value_.
+  void setLocalElement(hiop::hiopVector* x, local_ordinal_type i, real_type val)
+  {
+    x->copyFromDev();
+    real_type *xdat = x->local_data_host();
+    xdat[i] = val;
+    x->copyToDev();
+  }
+  /// Returns element _i_ of vector _x_.
+  real_type getLocalElement(const hiop::hiopVector* x, local_ordinal_type i)
+  {
+    x->copyFromDev();
+    return x->local_data_host_const()[i];
+  }
+
+  /// Returns pointer to local vector data
+  local_ordinal_type getLocalSize(const hiop::hiopVector* x)
+  {
+    return static_cast<local_ordinal_type>(x->get_local_size());
+  }
+
+  /// Returns pointer to local vector data
+  virtual real_type* getLocalData(hiop::hiopVector* x)
+  {
+    x->copyFromDev();
+    return x->local_data_host();
+  }
+
+  /// Checks if _local_ vector elements are set to `answer`.
+  int verifyAnswer(hiop::hiopVector* x, real_type answer)
+  {
+    x->copyFromDev();
+    const local_ordinal_type N = getLocalSize(x);
+    auto* xdata = static_cast<const real_type*>(getLocalData(x));
+    
+    int local_fail = 0;
+
+    for(local_ordinal_type i = 0; i < N; ++i)
+    {
+      if(!isEqual(xdata[i], answer))
```

Comment:
Not with `getElement` as is implemented now. Probably the best way to implement `verifyAnswer` method(s) is indeed to copy data to the host mirror and then perform all comparisons there. 

Furthermore, method `verifyAnswer` is supposed to be implementation specific and should not be part of `vectorTests.hpp`. Since its implementation is simple and straightforward, I am not sure if it is worthwhile to put too much effort in making it general.

Reply:
sgtm 👍